### PR TITLE
[system-probe] Add support for namespaces in conntrack code

### PR DESCRIPTION
**Pull request by hmahmood**
### What does this PR do?

This adds support for namespaces in the conntrack code in system-probe. Currently we are only tracking conntrack entries in the root namespace.

### Motivation

This is required for things like istio, which make nat table changes to the pod namespace. These resulting nat happening on the flows is not visible to us currently.

### CODE REVIEW

File: pkg/network/netlink/consumer.go [hunk 1]
```diff
@@ -128,19 +132,143 @@ func (c *Consumer) Events() <-chan Event {
...
+			Type: unix.RTM_GETNSID,
+			Sequence: netlinkSeqNumber,
+		},
+		Data: []byte{unix.AF_UNSPEC, 0, 0, 0},
+	}
+
+	msg.Data = append(msg.Data, data...)
+
+	if msg, err = conn.Send(msg); err != nil {
+		log.Tracef("isPeerNS: err sending netlink request: %s", err)
```
Comment (heyronhay) [inline pkg/network/netlink/consumer.go:162]: Why are these logged as trace and not something higher?

File: pkg/network/netlink/consumer.go [hunk 2]
```diff
@@ -128,19 +132,143 @@ func (c *Consumer) Events() <-chan Event {
...
+		return false
+	}
+
+	msgs, err := conn.Receive()
+	if err != nil {
+		log.Tracef("error receiving netlink reply: %s", err)
```
Comment (heyronhay) [inline pkg/network/netlink/consumer.go:168]: Ditto

File: pkg/network/netlink/consumer.go [hunk 3]
```diff
@@ -128,19 +132,143 @@ func (c *Consumer) Events() <-chan Event {
...
 		c.streaming = true
 		_ = c.conn.JoinGroup(netlinkCtNew)
-		c.receive(output)
+		c.receive(output, c.socket)
 	})
 
 	return output
 }
 
+func isPeerNS(conn *netlink.Conn, ns netns.NsHandle) bool {
```
Comment (heyronhay) [inline pkg/network/netlink/consumer.go:141]: Mind adding a bit of a comment?  Just the important stuff.

File: pkg/network/netlink/consumer.go [hunk 4]
```diff
@@ -128,19 +132,143 @@ func (c *Consumer) Events() <-chan Event {
...
 // This method is meant to be used once during the process initialization of system-probe.
 func (c *Consumer) DumpTable(family uint8) <-chan Event {
+	nss, err := util.GetNetNamespaces(c.procRoot)
+	if err != nil {
+		log.Errorf("could not get network namespaces: %s", err)
+	}
+
+	rootNS, err := netns.GetFromPath(fmt.Sprintf("%s/1/ns/net", c.procRoot))
+	if err != nil {
+		log.Errorf("could not get root namespace: %s", err)
```
Comment (heyronhay) [inline pkg/network/netlink/consumer.go:208]: Should you return here?

File: pkg/network/netlink/consumer.go [hunk 5]
```diff
@@ -128,19 +132,143 @@ func (c *Consumer) Events() <-chan Event {
...
+	return false
+}
+
 // DumpTable returns a channel of Event objects containing all entries
 // present in the Conntrack table. The channel is closed once all entries are read.
 // This method is meant to be used once during the process initialization of system-probe.
 func (c *Consumer) DumpTable(family uint8) <-chan Event {
+	nss, err := util.GetNetNamespaces(c.procRoot)
+	if err != nil {
+		log.Errorf("could not get network namespaces: %s", err)
```
Comment (heyronhay) [inline pkg/network/netlink/consumer.go:203]: Should you return here?

File: pkg/network/netlink/consumer.go [hunk 6]
```diff
@@ -128,19 +132,143 @@ func (c *Consumer) Events() <-chan Event {
...
+	defer func() {
+		if rootNS.IsOpen() {
+			rootNS.Close()
+		}
+	}()
+
+	conn, err := netlink.Dial(unix.AF_UNSPEC, &netlink.Config{NetNS: int(rootNS)})
+	if err != nil {
+		log.Errorf("could not open netlink socket: %s", err)
+	}
```
Comment (heyronhay) [inline pkg/network/netlink/consumer.go:220]: Should you return here?

File: pkg/process/util/netns.go [hunk 7]
```diff
@@ -13,28 +15,67 @@ import (
...
-	if err := netns.Set(rootNS); err != nil {
+	if err := netns.Set(ns); err != nil {
 		return err
 	}
 
 	fn()
 	return netns.Set(prevNS)
 }
+
+func GetNetNamespaces(procRoot string) ([]netns.NsHandle, error) {
```
Comment (heyronhay) [inline pkg/process/util/netns.go:48]: Function comment would be good!

**Review by heyronhay — CHANGES_REQUESTED** [review]
Initial pass - a few comments/requests!

File: pkg/network/netlink/consumer.go (hunk 1 above)
Comment (hmahmood) [reply pkg/network/netlink/consumer.go]: Don't know why I chose `Tracef`, these should be `Errorf`. Fixed

File: pkg/network/netlink/consumer.go (hunk 4 above)
Comment (hmahmood) [reply pkg/network/netlink/consumer.go]: Yes, see changes.

**Review by leeavital — APPROVED** [review]

**Review by heyronhay — APPROVED** [review]
Thanks for the changes!

**Pull request merged**